transform_inline.ml

```diff
@@ -0,0 +1,312 @@
+open Instr
+open Edit
+
+(* Unlike `analysis_input`, this preserves the order of `formals`. This
+   facilitates matching the formals to the arguments passed at a callsite.*)
+type inlining_input = {
+  formals : variable list;
+  instrs : instructions;
+}
+
+let as_inlining_input (func:afunction) (version:version) : inlining_input =
+  { formals = Analysis.as_var_list func.formals; instrs = version.instrs }
+
+(* FUNCTION INLINING *)
+(* Given a program, inline the functions in it to the maximum possible extent.
+   Transitively recursive calls are inlined until they become self recursive.
+   Self recursive callsites are not inlined. Intermediate inlined versions are
+   stored in the caller. *)
+let inline ({main; functions} as orig_prog : program) : program option =
+
+  (* Given the formals and instructions of a function, get all the declared and
+     used vars in the function. *)
+  let function_vars {formals; instrs} =
+    let all_declared_vars instrs = Array.fold_left
+        (fun vars instr -> VarSet.union vars (declared_vars instr))
+        VarSet.empty
+        instrs
+    in
+    (* This union is needed in case a formal is not used in the instructions. *)
+    VarSet.union (all_declared_vars instrs) (VarSet.of_list formals)
+  in
+
+  (* Given a function and an array of instructions, generate a fresh version
+     for the function with these instructions. *)
+  let add_version (func : afunction) (instrs : instructions)  =
+    let label = fresh_version_label func "inlined_version" in
+    let version = {instrs = instrs; label = label; annotations = None} in
+    {func with body = List.cons version func.body}
+  in
+
+  (* Given a list of (old_name, new_name) pairs, replace all occurrences of an
+     `old_name` with the `new_name` in the instruction. Note that this process
+     has an inherent subtlety. Consider vars_alist = [(x, x_1); (x_1, x_2)].
+     All occurrences of x have to be replaced by x_1 and all occurrences of x_1
+     have to be replaced by x_2. But an occurrence of x replaced by x_1 should
+     not again be replaced by x_2. Hence the function works on the entire list
+     of replacement pairs at once rather than being called on one pair at a
+     time. *)
+  (* TODO - fix using visitor pattern if appropriate.*)
```

Comment:
It seems @gasche 's `Instr.map` should be used here. See `Edit.replace_uses` for an example on how to use it. The idea is that you inherit from `Instr.map` and override all the functions that correspond to the parts of instructions you want to change.